extensions/ql-vscode/src/queries-panel/query-discovery.ts

```diff
@@ -1,154 +1,116 @@
-    for (const workspaceFolder of workspaceFolders) {
-      const root = await this.discoverQueriesInWorkspace(workspaceFolder);
-      if (root !== undefined) {
-        rootDirectories.push(root);
+  public buildQueryTree(): Array<FileTreeNode<string>> {
+    const roots = [];
+    for (const workspaceFolder of getOnDiskWorkspaceFoldersObjects()) {
+      const queriesInRoot = this.paths.filter((query) =>
+        containsPath(workspaceFolder.uri.fsPath, query.path),
+      );
+      if (queriesInRoot.length > 0) {
+        const root = new FileTreeDirectory<string>(
+          workspaceFolder.uri.fsPath,
+          workspaceFolder.name,
+          this.env,
+        );
+        for (const query of queriesInRoot) {
+          const dirName = dirname(normalize(relative(root.path, query.path)));
+          const parentDirectory = root.createDirectory(dirName);
+          parentDirectory.addChild(
+            new FileTreeLeaf<string>(
+              query.path,
+              basename(query.path),
+              query.language,
+            ),
+          );
+        }
+        root.finish();
+        roots.push(root);
       }
     }
-    return rootDirectories;
+    return roots;
   }
 
-  private async discoverQueriesInWorkspace(
-    workspaceFolder: WorkspaceFolder,
-  ): Promise<FileTreeDirectory<string> | undefined> {
-    const fullPath = workspaceFolder.uri.fsPath;
-    const name = workspaceFolder.name;
+  protected async getDataForPath(path: string): Promise<Query> {
+    const language = this.determineQueryLanguage(path);
+    return { path, language };
+  }
 
-    // We don't want to log each invocation of resolveQueries, since it clutters up the log.
-    const silent = true;
-    const resolvedQueries = await this.cliServer.resolveQueries(
-      fullPath,
-      silent,
-    );
-    if (resolvedQueries.length === 0) {
-      return undefined;
-    }
+  protected pathIsRelevant(path: string): boolean {
+    return path.endsWith(QUERY_FILE_EXTENSION);
+  }
 
-    const rootDirectory = new FileTreeDirectory<string>(
-      fullPath,
-      name,
-      this.env,
-    );
-    for (const queryPath of resolvedQueries) {
-      const relativePath = normalize(relative(fullPath, queryPath));
-      const dirName = dirname(relativePath);
-      const parentDirectory = rootDirectory.createDirectory(dirName);
-      parentDirectory.addChild(
-        new FileTreeLeaf<string>(queryPath, basename(queryPath), "language"),
-      );
+  protected shouldOverwriteExistingData(
+    newData: Query,
+    existingData: Query,
+  ): boolean {
+    return newData.language !== existingData.language;
+  }
+
+  private recomputeAllQueryLanguages() {
+    // All we know is that something has changed in the set of known query packs.
+    // We have no choice but to recompute the language for all queries.
+    for (const query of this.paths) {
+      query.language = this.determineQueryLanguage(query.path);
     }
+    this.onDidChangePathsEmitter.fire();
```

Comment:
It seems strange to be to be re-using this when none of the paths actually changed. Perhaps we should be thinking about composition rather than inheritance?

Reply:
(Using this thread to talk about the idea of composition vs inheritance)

I don't think composition works here, or at least it doesn't make things easier. Or possibly I'm not understanding what you're suggesting. Could you give a little more info about what you had in mind?

The main issue I'm hitting is that we aren't just tracking paths here but we're also tracking some extra data about each path, such as the language of the query pack. We want to avoid re-computing this extra data when not needed, so we want the computation of the data to be tightly integrated so we only do it when needed, so it makes sense for it to be done inside `FilePathDiscovery`.

So essentially I don't think we can remove all customisation from `FilePathDiscovery`. It could have customisation by abstract methods or function closures passed to the constructor, and either works but I think abstract methods is a bit cleaner.

We could essentially split `QueryDiscovery` into two, so there's one class that is the public interface and has `onDidChangeQueries` and `buildQueryTree`, and that class owns a `QueryFilePathDiscovery extends FilePathDiscovery`. Is this what you meant? It's possible but I don't see any great benefits and it makes accessing fields on other classes more clunky.

---

I wonder perhaps if the real problem is the naming of the fields on `FileTreeDiscovery`? It's not just tracking paths but also data about those paths, so then it sounds odd to just say "the paths have changed". I could rename everything from just "path" to "pathData" or another name you think it clearer?

Reply:
I didn't have a specific idea for the composition, but I think it might be helpful. I agree that we'd just be trading abstract methods for callbacks passed into the constructor, so it might no make sense here.

Perhaps instead we should have a separate event emitter from the `onDidChangePathsEmitter` emitter? We could then do something like this int he constructor:

```ts
    this.push(
      this.onDidChangePathsEmitter.event(() =>
        this.onDidChangeQueriesEmitter.fire(),
      ),
    );
```

The `onDidChangePathsEmitter` could then be made private and only `onDidChangePathsEmitter.event` could be exposed (probably `protected`). This would at least make the separation between the `FilePathDiscovery` and the `QueryDiscovery` clear.

I think changing the naming of the fields would also help by clarifying that it's data about paths, and not just the paths itself. I'm not sure if you'd still need the separation between the two event emitters in that case. However, I do think we are leaking some implementation details of the `FilePathDiscovery` by directly modifying `this.paths`. Could we instead make a method on `FilePathDiscovery` that allows you to update all path data? Something like this:

```ts
  protected async updateAllPathData(
    callback: (pathData: PathData) => Promise<void>,
  ): Promise<void> {
    await Promise.all(this.paths.map(callback));
    this.onDidChangePathsEmitter.fire();
  }
```

Then the subclass doesn't need to know anything about the event emitter and can just call that method. The `onDidChangePathsEmitter` can be private in that case, and only `onDidChangePathsEmitter.event` can be exposed to subclasses.

Reply:
I'll update the name to be `pathData` throughout so it's more clear what's what.

I do very much like the idea of making `pathData` private to `FilePathDiscovery` and adding a `updateAllPathData` method to do that operation. I'll implement these two bits.

At this point I'm not sure we need a second event emitter. Unless it makes things noticeably clearer I'd prefer to stick to just one.

I'm still unsure about going fully composable. I've played around with it a bit, and I think there is potential there but it's also a lot of work to get to a point where you see the benefits. I'd prefer to stay with the current approach for now if that's ok. It doesn't block us from changing approach in the future, and we do have test coverage right now. We could make the interfaces cleaner and make things a bit easier to test but it's a lot of work rearranging things to get there.